Kernel/Heap/kmalloc.cpp

```diff
@@ -182,17 +183,67 @@ struct KmallocGlobalHeap {
     }
 };
 
+[[gnu::malloc, gnu::returns_nonnull, gnu::alloc_size(1)]] void* kmalloc_static_eternal(size_t);
+
+class EternalHeap {
+    AK_MAKE_NONCOPYABLE(EternalHeap);
+    AK_MAKE_NONMOVABLE(EternalHeap);
+
+public:
+    void* operator new(size_t size) { return kmalloc_static_eternal(size); }
+
+public:
+    EternalHeap(u8* memory, size_t memory_size)
+        : m_memory_start(memory)
+        , m_memory_size(memory_size)
+    {
+    }
+    EternalHeap(NonnullOwnPtr<Region>&& responsible_region)
+        : m_responsible_memory_region(move(responsible_region))
+        , m_memory_start(responsible_region->vaddr().as_ptr())
+        , m_memory_size(responsible_region->size())
+    {
+    }
+    ~EternalHeap() { VERIFY_NOT_REACHED(); }
+
+    void* allocate(size_t size)
+    {
+        size = round_up_to_power_of_two(size, sizeof(void*));
+        if (free_bytes() < size)
+            return nullptr;
+        void* ptr = VirtualAddress(m_memory_start).offset(m_allocated_bytes_count).as_ptr();
+        m_allocated_bytes_count += size;
+        return ptr;
+    }
+
+    const u8* memory() const { return m_memory_start; }
+    size_t total_bytes() const { return m_memory_size; }
+    size_t free_bytes() const { return m_memory_size - m_allocated_bytes_count; }
+    size_t allocated_bytes() const { return m_allocated_bytes_count; }
+
+private:
+    OwnPtr<Region> m_responsible_memory_region;
+    const u8* m_memory_start;
+    const size_t m_memory_size;
+    size_t m_allocated_bytes_count { 0 };
+};
+
+// Note: We ensure we are aligned because kmalloc_static_eternal() does the same.
+static_assert(round_up_to_power_of_two(sizeof(EternalHeap), sizeof(void*)) == sizeof(EternalHeap));
+
 READONLY_AFTER_INIT static KmallocGlobalHeap* g_kmalloc_global;
 alignas(KmallocGlobalHeap) static u8 g_kmalloc_global_heap[sizeof(KmallocGlobalHeap)];
 
 // Treat the heap as logically separate from .bss
-__attribute__((section(".heap"))) static u8 kmalloc_eternal_heap[ETERNAL_RANGE_SIZE];
+__attribute__((section(".heap"))) static u8 kmalloc_eternal_heaps[ETERNAL_RANGE_SIZE];
```

Comment:
Hmm, this reserves 3 MiB of BSS just to hold the very small EternalHeap objects? That seems like a huge waste of memory :)

Reply:
Yeah, maybe 100 KiB will be enough.